common/src/main/java/com/wynntils/models/emeralds/EmeraldModel.java

```diff
@@ -50,6 +50,93 @@ public EmeraldModel(CharacterModel characterModel, ItemModel itemModel) {
         super(List.of(characterModel, itemModel));
     }
 
+    @SubscribeEvent
+    public void onWorldChange(WorldStateEvent e) {
+        if (e.getNewState() != WorldState.WORLD) return;
+
+        inventoryEmeralds = 0;
+        containerEmeralds = 0;
+
+        // Rescan inventory at login
+        Inventory inventory = McUtils.inventory();
+        for (int i = 0; i < inventory.getContainerSize(); i++) {
+            adjustBalance(null, inventory.getItem(i), true);
+        }
+    }
+
+    @SubscribeEvent
+    public void onSetSlot(SetSlotEvent.Post event) {
+        boolean isInventory = event.getContainer() == McUtils.inventory();
+        if (pouchContainerId != -1 && !isInventory) return;
+
+        // FIXME: This is a hack to always have up-to-date emerald counts
+        //        When Wynncraft fixes emerald stacking,
+        //        this can be simplified greatly (by using old and new stacks)
+        //        However, this is really fast so maybe we can keep it (pending profiling)
+        if (isInventory) {
+            inventoryEmeralds = 0;
+
+            // Rescan inventory after merging items
+            List<ItemStack> items = McUtils.inventoryMenu().getItems();
+            for (ItemStack item : items) {
+                adjustBalance(null, item, true);
+            }
+        } else if (event.getContainer() == McUtils.containerMenu()) {
+            containerEmeralds = 0;
+
+            // Rescan container after merging items
+            List<ItemStack> items = McUtils.containerMenu().getItems();
+            for (ItemStack item : items) {
+                adjustBalance(null, item, false);
+            }
+        }
+    }
+
+    @SubscribeEvent
+    public void onMenuOpened(MenuEvent.MenuOpenedEvent e) {
+        String title = WynnUtils.normalizeBadString(e.getTitle().getString());
+        if (title.equals("Emerald Pouch")) {
```

Comment:
Worth making this a constant? Think we do that in most other places

Reply:
Perhaps but this is a move, not a change.